### PR TITLE
Add KVMap for MQTTContext and update Kafka filter

### PR DESCRIPTION
**Pull request by suchen-sci**

### CODE REVIEW

File: pkg/filter/kafka/spec.go [hunk 1]
```diff
@@ -21,5 +21,13 @@ type (
 	// Spec is spec of Kafka
 	Spec struct {
 		Backend []string `yaml:"backend" jsonschema:"required,uniqueItems=true"`
+		KVMap   *KVMap   `yaml:"mqtt" jsonschema:"required"`
```
Comment (zhao-kun) [inline pkg/filter/kafka/spec.go:24]: Kafka spec could contain topic (string) and headers (map[string]string) directly. It no need to introduce a sub struct `KVMap`

File: pkg/filter/kafka/spec.go [hunk 2]
```diff
@@ -21,5 +21,13 @@ type (
...
+	}
+
+	// KVMap defines ways to get kafka data from MQTTContext kv map
+	KVMap struct {
+		Topic   string `yaml:"topic" jsonschema:"required"`
+		Headers string `yaml:"headers" jsonschema:"required"`
+		Payload string `yaml:"payload" jsonschema:"required"`
```
Comment (zhao-kun) [inline pkg/filter/kafka/spec.go:31]: Default payload has no sense, remove it

File: pkg/filter/kafka/kafka.go [hunk 3]
```diff
@@ -125,22 +141,55 @@ func (k *Kafka) Status() interface{} {
...
-	if ctx.PacketType() != context.MQTTPublish {
-		return &context.MQTTResult{}
+	var topic string
+	var headers map[string]string
+	var payload []byte
+	var ok bool
+
+	// set data from kv map
+	if k.topic != "" {
+		topic, ok = ctx.GetKV(k.topic).(string)
```
Comment (zhao-kun) [inline pkg/filter/kafka/kafka.go:151]: The topic value in the `Context` has higher priority then default topic.

File: pkg/filter/kafka/kafka.go [hunk 4]
```diff
@@ -125,22 +141,55 @@ func (k *Kafka) Status() interface{} {
...
+		if !ok {
+			return &context.MQTTResult{ErrString: resultGetDataFailed}
+		}
+	}
+	if k.headers != "" {
```
Comment (zhao-kun) [inline pkg/filter/kafka/kafka.go:156]: The headers value in the `Context` has higher priority then default headers

File: pkg/filter/kafka/kafka.go [hunk 5]
```diff
@@ -125,22 +141,55 @@ func (k *Kafka) Status() interface{} {
...
+	if k.payload != "" {
+		payload, ok = ctx.GetKV(k.payload).([]byte)
+		if !ok {
+			return &context.MQTTResult{ErrString: resultGetDataFailed}
+		}
 	}
 
-	p := ctx.PublishPacket()
-	logger.Debugf("produce msg with topic %s", p.Topic())
+	// set data from PublishPacket if data is missing
```
Comment (zhao-kun) [inline pkg/filter/kafka/kafka.go:169]: Payload in the Context is higher priority then PublishPacket

File: pkg/filter/kafka/spec.go (hunk 1 above)
Comment (suchen-sci) [reply pkg/filter/kafka/spec.go:24]: `KVMap` defines keys to get values from MQTT context KVMap. For example, if set KVMap.Topic = "kafkaTopic", then topic for kafka backend will be MQTTContext.GetKV("kafkaTopic"). maybe i use wrong name.

File: pkg/filter/kafka/spec.go (hunk 2 above)
Comment (suchen-sci) [reply pkg/filter/kafka/spec.go:31]: If not use `Payload`, then we only can send data to kafka backend by using publish packet. For example, some mqtt client disconnect, and i want to send some data to kafak, then i can use MQTTContext.SetKV("payload", "some data") and use MQTTContext.GetKV("payload") to get it. `Payload` here is the key to call GetKV...

Comment (zhao-kun) [reply pkg/filter/kafka/spec.go:31]: Sorry for my misunderstood. The implementation is OK

File: pkg/filter/kafka/kafka.go (hunk 5 above)
Comment (suchen-sci) [reply pkg/filter/kafka/kafka.go:173]: updated!

**Review by localvar — APPROVED** [review]

**Pull request merged**